### PR TITLE
Allow random starting location in scenarios

### PR DESCRIPTION
**Pull request by ZhilkinSerg**
#### Summary

```SUMMARY: Interface "Allow random starting location in scenarios"```

#### Purpose of change

Allow random starting location in scenarios.

#### Describe the solution

Add entry to starting location selection menu (menu selection defaults to this entry). When selected it allows starting in random location (from the list of starting locations available in scenario) instead of starting in location explicitly selected in menu.

#### Testing

- [ ] - Start a game using "* Random location *" menu entry and test that player character is put to different random starting locations (from the list of scenario starting locations)

- [ ] - Start a game NOT using "* Random location *" menu entry and test that player character is put to starting location selected in starting location selection menu

- [ ] - Test loading of legacy and current character templates

#### Additional context

State of random starting location selection is saved in character template files (when reading legacy templates fallback to non-random start locations).

### CODE REVIEW

File: src/newcharacter.cpp [hunk 1]
```diff
@@ -83,6 +83,9 @@
 
 #define NEWCHAR_TAB_MAX 6 // The ID of the rightmost tab
 
+static constexpr int RANDOM_START_LOCATION_ENTRY = INT_MIN;
+static const std::string RANDOM_START_LOCATION_TEXT = "<color_red>* Random location *</color>";
```
Comment (AMurkin) [inline src/newcharacter.cpp]: Not `translation` or `translate_marker()`?

File: src/newcharacter.cpp [hunk 2]
```diff
@@ -2627,11 +2644,13 @@ bool avatar::load_template( const std::string &template_name, points_left &point
             points.skill_points = jobj.get_int( "skill_points" );
             points.limit = static_cast<points_left::point_limit>( jobj.get_int( "limit" ) );
 
+            this->random_start_location = jobj.get_bool( "random_start_location", true );
```
Comment (BevapDin) [inline src/newcharacter.cpp]:
```suggestion
            random_start_location = jobj.get_bool( "random_start_location", true );
```

File: src/newcharacter.cpp [hunk 3]
```diff
@@ -2627,11 +2644,13 @@ bool avatar::load_template( const std::string &template_name, points_left &point
...
             const std::string jobj_start_location = jobj.get_string( "start_location", "" );
 
             // g->scen->allowed_start( loc.ident() ) is checked once scenario loads in avatar::load()
             for( const auto &loc : start_location::get_all() ) {
                 if( loc.ident().str() == jobj_start_location ) {
+                    this->random_start_location = false;
```
Comment (BevapDin) [inline src/newcharacter.cpp]:
```suggestion
                    random_start_location = false;
```

File: src/newcharacter.cpp [hunk 4]
```diff
@@ -2177,19 +2180,26 @@ tab_direction set_description( const catacurses::window &w, avatar &you, const b
 
     uilist select_location;
     select_location.text = _( "Select a starting location." );
-    int offset = 0;
+    int offset = 1;
+    uilist_entry entry_random_start_location( RANDOM_START_LOCATION_ENTRY, true, -1,
```
Comment (BevapDin) [inline src/newcharacter.cpp]: Considering that `RANDOM_START_LOCATION_ENTRY` is ony used in this function, you should just make it a local variable.
```C++
    const std::string RANDOM_START_LOCATION_TEXT = _( "<color_red>* Random location *</color>" );
```

File: src/newcharacter.cpp [hunk 5]
```diff
@@ -2627,12 +2644,14 @@ bool avatar::load_template( const std::string &template_name, points_left &point
...
 
+            random_start_location = jobj.get_bool( "random_start_location", true );
             const std::string jobj_start_location = jobj.get_string( "start_location", "" );
 
             // g->scen->allowed_start( loc.ident() ) is checked once scenario loads in avatar::load()
             for( const auto &loc : start_location::get_all() ) {
                 if( loc.ident().str() == jobj_start_location ) {
-                    this->start_location = loc.ident();
+                    random_start_location = false;
+                    start_location = loc.ident();
```
Comment (ZhilkinSerg) [inline src/newcharacter.cpp]:
```suggestion
                    this->start_location = loc.ident();
```

**Pull request merged**